File: packages/sdk-typescript/src/guard.ts
/** The `AgentGuard` facade — identity, tool interception, enforcement (PRD §37). */

import { randomUUID } from "node:crypto";

import { AgentGuardClient, type FetchLike } from "./client.js";
import {
  type AgentGuardConfig,
  type ConfigOverrides,
  resolveConfig,
} from "./config.js";
import {
  ApprovalRequired,
  ConfigurationError,
  PolicyDenied,
  RateLimited,
  RuntimeUnavailable,
} from "./errors.js";
import { applyRedactions } from "./redact.js";
import { Decision, type DecisionResult, unavailableResult } from "./types.js";

export interface AgentGuardOptions extends ConfigOverrides {
  /** Inject a fetch implementation (tests, proxies). */
  fetch?: FetchLike;
  /** Register the agent automatically on first use. Default true. */
  autoRegister?: boolean;
}

export interface EvaluateOptions {
  action?: string;
  context?: Record<string, unknown>;
  requestId?: string;
  dataClassification?: string;
}

type Params = Record<string, unknown>;
// deliberately permissive: the wrapper forwards whatever the caller passes
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFn = (...args: any[]) => unknown;

export class AgentGuard {
  readonly config: AgentGuardConfig;
  readonly client: AgentGuardClient;
  private agentIdCache: string | null = null;

  constructor(options: AgentGuardOptions = {}) {
    this.config = resolveConfig(options);
    this.client = new AgentGuardClient({
      apiKey: this.config.apiKey,
      baseUrl: this.config.baseUrl,
      timeoutMs: this.config.timeoutMs,
      fetch: options.fetch,
    });
  }

  /** Resolve (and, if needed, register) the agent id for the configured agent. */
  async agentId(): Promise<string> {
    if (this.agentIdCache === null) {
      if (!this.config.agent) {
        throw new ConfigurationError(
          "no agent — pass agent to AgentGuard() or set AGENTGUARD_AGENT",
        );
      }
      this.agentIdCache = await this.client.resolveAgentId({
        name: this.config.agent,
        environment: this.config.environment,
      });
    }
    return this.agentIdCache;
  }

  /** Pre-set the agent id (skips the identity lookup). */
  setAgentId(id: string): void {
    this.agentIdCache = id;
  }

  /** Ask the runtime what to do. Returns a decision; only throws
   * `RuntimeUnavailable` when fail_mode='closed' and the API is unreachable. */
  async evaluate(
    tool: string,
    parameters: Params = {},
    options: EvaluateOptions = {},
  ): Promise<DecisionResult> {
    const requestId = options.requestId ?? randomUUID();
    try {
      return await this.client.evaluate({
        agentId: await this.agentId(),
        tool,
        action: options.action ?? "execute",
        parameters,
        context: options.context ?? {},
        requestId,
        dataClassification: options.dataClassification,
      });
    } catch (err) {
      if (err instanceof RuntimeUnavailable) {
        if (this.config.failMode === "open") {
          const result = unavailableResult("fail-open");
          return {
            ...result,
            decision: Decision.Allow,
            riskScore: 0,
            riskSeverity: "info",
            reasons: ["runtime unavailable — fail-open"],
            failMode: "fail_open",
            allowed: true,
            requestId,
          };
        }
      }
      throw err;
    }
  }

  /** Evaluate and enforce. Resolves to `{ result, parameters }` on ALLOW /
   * REDACT (parameters may be a masked copy); rejects otherwise. */
  async check(
    tool: string,
    parameters: Params = {},
    options: EvaluateOptions = {},
  ): Promise<{ result: DecisionResult; parameters: Params }> {
    let result: DecisionResult;
    try {
      result = await this.evaluate(tool, parameters, options);
    } catch (err) {
      if (err instanceof RuntimeUnavailable) {
        throw new PolicyDenied(
          `runtime unavailable (fail-closed): ${err.message}`,
          unavailableResult(err.message),
        );
      }
      throw err;
    }

    switch (result.decision) {
      case Decision.Allow:
        return { result, parameters };
      case Decision.Redact:
        return { result, parameters: applyRedactions(parameters, result.redactions) };
      case Decision.Approval:
        throw new ApprovalRequired(result.reasons.join("; ") || "approval required", result);
      case Decision.RateLimit:
        throw new RateLimited(result.reasons.join("; ") || "rate limited", result);
      default:
        throw new PolicyDenied(result.reasons.join("; ") || "denied", result);
    }
  }

  /** Wrap a tool function so every call is enforced. The wrapped function is
   * async. It expects a single plain-object argument of named parameters
   * (the shape every major agent framework uses); on REDACT the masked copy is
   * passed through. Positional-arg tools are evaluated but not redacted. */
  tool<F extends AnyFn>(fn: F, options: { name?: string; action?: string } = {}): F {
    const toolName = options.name ?? (fn.name || "tool");
    const guard = this;
    const wrapped = async function (this: unknown, ...args: unknown[]) {
      const single =
        args.length === 1 && args[0] != null && typeof args[0] === "object" && !Array.isArray(args[0]);
      const params: Params = single ? (args[0] as Params) : { args };
      const { result, parameters } = await guard.check(toolName, params, {
        action: options.action,
      });
      const callArgs = single && result.decision === Decision.Redact ? [parameters] : args;
      return fn.apply(this, callArgs);
    };
    Object.defineProperty(wrapped, "name", { value: toolName });
    (wrapped as { agentGuardTool?: string }).agentGuardTool = toolName;
    return wrapped as unknown as F;
  }

  /** Register the agent and, if `target.tools` is a discoverable array, wrap it. */
  async protect<T extends { tools?: unknown[] }>(target: T): Promise<T> {
    await this.agentId();
    if (Array.isArray(target.tools)) {
      target.tools = target.tools.map((t) => (typeof t === "function" ? this.tool(t as AnyFn) : t));
    }
    (target as { agentguard?: AgentGuard }).agentguard = this;
    return target;
  }

  /** Block until an approval request is decided. Returns its final status. */
  async waitForApproval(
    approvalRequestId: string,
    options: { pollMs?: number; timeoutMs?: number } = {},
  ): Promise<string> {
    const pollMs = options.pollMs ?? 2000;
    const deadline = Date.now() + (options.timeoutMs ?? 300_000);
    while (Date.now() < deadline) {
      const row = (await this.client.get(`/v1/approvals/${approvalRequestId}`)) as {
        status: string;
      };
      if (row.status !== "pending") return row.status;
      await new Promise((r) => setTimeout(r, pollMs));
    }
    return "timeout";
  }
}

File: packages/sdk-python/agentguard/guard.py
"""The `AgentGuard` facade — identity, tool interception, enforcement (PRD §37)."""

from __future__ import annotations

import functools
import inspect
import time
import uuid
from collections.abc import Callable
from typing import Any, TypeVar

import httpx

from . import config as _config
from .client import Client
from .decision import Decision, DecisionResult
from .exceptions import (
    ApprovalRequired,
    ConfigurationError,
    PolicyDenied,
    RateLimited,
    RuntimeUnavailable,
)
from .redact import redact_params

F = TypeVar("F", bound=Callable[..., Any])


class AgentGuard:
    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        agent: str | None = None,
        environment: str | None = None,
        fail_mode: str | None = None,
        timeout: float | None = None,
        session: httpx.Client | None = None,
        auto_register: bool = True,
    ) -> None:
        self.config = _config.resolve(
            api_key=api_key,
            base_url=base_url,
            agent=agent,
            environment=environment,
            fail_mode=fail_mode,
            timeout=timeout,
        )
        self._client = Client(
            api_key=self.config.api_key,
            base_url=self.config.base_url,
            timeout=self.config.timeout,
            session=session,
        )
        self._auto_register = auto_register
        self._agent_id: str | None = None

    # --- lifecycle -------------------------------------------------------

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> AgentGuard:
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    @property
    def client(self) -> Client:
        return self._client

    # --- identity -----------------------------------------------------

    @property
    def agent_id(self) -> str:
        if self._agent_id is None:
            if not self.config.agent:
                raise ConfigurationError(
                    "no agent — pass agent= to AgentGuard() or set AGENTGUARD_AGENT"
                )
            self._agent_id = self._client.resolve_agent_id(
                name=self.config.agent, environment=self.config.environment
            )
        return self._agent_id

    def protect(self, agent: Any = None, *, tools: list[Callable] | None = None) -> Any:
        """Register the agent and, if discoverable, wrap its tools (PRD §38)."""
        _ = self.agent_id  # force identity resolution / registration
        candidates = tools
        if candidates is None and agent is not None:
            candidates = list(getattr(agent, "tools", []) or [])
        if candidates:
            wrapped = [self.tool(fn) if callable(fn) else fn for fn in candidates]
            if agent is not None and hasattr(agent, "tools"):
                try:
                    agent.tools = wrapped
                except (AttributeError, TypeError):
                    pass
        if agent is not None:
            agent.agentguard = self  # type: ignore[attr-defined]
        return agent

    # --- evaluation -------------------------------------------------

    def evaluate(
        self,
        tool: str,
        parameters: dict | None = None,
        *,
        action: str = "execute",
        context: dict | None = None,
        request_id: str | None = None,
        data_classification: str | None = None,
    ) -> DecisionResult:
        """Ask the runtime what to do. Never raises for a normal decision;
        raises ``RuntimeUnavailable`` only when fail_mode='closed' and the API
        is unreachable."""
        rid = request_id or uuid.uuid4().hex
        try:
            return self._client.evaluate(
                agent_id=self.agent_id,
                tool=tool,
                action=action,
                parameters=parameters or {},
                context=context or {},
                request_id=rid,
                data_classification=data_classification,
            )
        except RuntimeUnavailable:
            if self.config.fail_mode == "open":
                return DecisionResult(
                    decision=Decision.ALLOW,
                    risk_score=0,
                    risk_severity="info",
                    request_id=rid,
                    reasons=["runtime unavailable — fail-open"],
                    fail_mode="fail_open",
                )
            raise

    def check(
        self,
        tool: str,
        parameters: dict | None = None,
        *,
        action: str = "execute",
        context: dict | None = None,
        request_id: str | None = None,
    ) -> tuple[DecisionResult, dict]:
        """Evaluate and enforce. Returns ``(result, effective_parameters)`` on
        ALLOW / REDACT; raises otherwise."""
        params = parameters or {}
        try:
            result = self.evaluate(
                tool, params, action=action, context=context, request_id=request_id
            )
        except RuntimeUnavailable as exc:
            raise PolicyDenied(
                f"runtime unavailable (fail-closed): {exc}", _unavailable(exc)
            ) from exc

        if result.decision == Decision.ALLOW:
            return result, params
        if result.decision == Decision.REDACT:
            return result, redact_params(params, result.redactions)
        if result.decision == Decision.APPROVAL:
            raise ApprovalRequired("; ".join(result.reasons) or "approval required", result)
        if result.decision == Decision.RATE_LIMIT:
            raise RateLimited("; ".join(result.reasons) or "rate limited", result)
        raise PolicyDenied("; ".join(result.reasons) or "denied", result)

    # --- tool decorator ---------------------------------------------

    def tool(
        self,
        func: F | None = None,
        *,
        name: str | None = None,
        action: str = "execute",
    ) -> F | Callable[[F], F]:
        def decorate(fn: F) -> F:
            tool_name = name or fn.__name__
            sig = inspect.signature(fn)

            @functools.wraps(fn)
            def wrapper(*args: Any, **kwargs: Any) -> Any:
                bound = sig.bind_partial(*args, **kwargs)
                params = _serialisable(dict(bound.arguments))
                result, effective = self.check(tool_name, params, action=action)
                if result.decision == Decision.REDACT:
                    new_bound = sig.bind_partial(**effective)
                    return fn(*new_bound.args, **new_bound.kwargs)
                return fn(*args, **kwargs)

            wrapper.agentguard_tool = tool_name  # type: ignore[attr-defined]
            return wrapper  # type: ignore[return-value]

        return decorate if func is None else decorate(func)

    # --- approvals -------------------------------------------------

    def wait_for_approval(
        self, approval_request_id: str, *, poll_seconds: float = 2.0, timeout: float = 300.0
    ) -> str:
        """Block until an approval is decided. Returns its final status."""
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            row = self._client.get(f"/v1/approvals/{approval_request_id}")
            if row["status"] != "pending":
                return row["status"]
            time.sleep(poll_seconds)
        return "timeout"


def _serialisable(arguments: dict) -> dict:
    out: dict = {}
    for k, v in arguments.items():
        if isinstance(v, (str, int, float, bool, type(None), list, dict)):
            out[k] = v
        else:
            out[k] = repr(v)
    return out


def _unavailable(exc: Exception) -> DecisionResult:
    return DecisionResult(
        decision=Decision.DENY,
        risk_score=100,
        risk_severity="critical",
        request_id="",
        reasons=[f"runtime unavailable: {exc}"],
        fail_mode="fail_closed",
    )
